bintray-tidy.sh

```diff
@@ -0,0 +1,256 @@
+  -s, --simulate        Show what would be deleted without actually deleting it.
+
+ENVIRONMENT:
+  * BINTRAY_USER, BINTRAY_API_KEY - required
+  * BINTRAY_REPO, BINTRAY_REPO_OWNER - optional
+
+EXAMPLES:
+
+ 1) Keep the 10 most recent versions and delete everything else:
+
+    $  ./bintray-tidy.sh  max-versions  10  MyRepo/CoolPackage
+
+ 2) Simulate running the archive strategy to see what would be deleted:
+
+    $  ./bintray-tidy.sh  -s  archive  DreamTeam/TeamRepo/SharedPackage
+
+                            ~~~ bintray-tidy.sh ~~~
+EOF
+}
+
+main() {
+  initialize_variables
+
+  echo "$0"
+  echo "Tidying package '${BINTRAY_REPO_OWNER}/${BINTRAY_REPO}/${package}'."
+
+  get_remote_versions
+
+  [ "$arg_s" ] && echo "SIMULATION: No versions will actually be deleted from the server." || true
+
+  case "${strategy}" in
+    max-versions )
+      max_versions
+      ;;
+    max-days )
+      max_days
+      ;;
+    archive )
+      archive
+      ;;
+    * )
+      fatal_error "invalid strategy '${strategy}'"
+      exit 1
+      ;;
+  esac
+
+  [ "$arg_s" ] && echo "SIMULATION: Nothing was deleted. Rerun without '-s' to actually delete versions." || true
+}
+
+function initialize_variables() {
+  # Constants
+  API="https://api.bintray.com"
+
+  # Required environment variables
+  BINTRAY_USER="${BINTRAY_USER:?Environment variable missing/empty!}" # env
+  BINTRAY_API_KEY="${BINTRAY_API_KEY:?Environment variable missing/empty!}" # env
+
+  # Optional environment variables
+  BINTRAY_REPO_OWNER="${BINTRAY_REPO_OWNER:-$BINTRAY_USER}" # env, or use BINTRAY_USER
+  BINTRAY_REPO="${BINTRAY_REPO:?Not specified and not set in environment.}" # env or CLI
+
+  # Variables
+  PCK_URL="${API}/packages/${BINTRAY_REPO_OWNER}/${BINTRAY_REPO}/${package}"
+  CURL="curl -u${BINTRAY_USER}:${BINTRAY_API_KEY} \
+         -H Content-Type:application/json \
+         -H Accept:application/json"
+}
+
+function get_remote_versions() {
+  readarray -t REMOTE_VERSIONS < <( \
+  curl -X GET "${PCK_URL}" \
+    | sed -nr 's|.*"versions":\["([^]]*)"\].*|\1|p' \
+    | sed 's|","|\n|g' )
+  echo "$0: ${#REMOTE_VERSIONS[@]} versions found on the remote server."
+}
+
+function max_versions() {
+  # Enforce upper limit on the number of versions:
+  for ((i=0;i<${#REMOTE_VERSIONS[@]};i++)); do
+    if [ "$i" -lt "$max_versions" ]; then
+      print_action "Keeping" "${REMOTE_VERSIONS[$i]}"
+    else
+      delete_version "${REMOTE_VERSIONS[$i]}" "$created"
+    fi
+  done
+}
+
+function max_days() {
+  # Delete versions that are more than ${max_days} days old:
+  cuttoff_date_YMD="$(date --utc -d "$max_days days ago" +%Y%m%d)" # YYYYMMDD
+  for ((i=0;i<${#REMOTE_VERSIONS[@]};i++)); do
+    get_version_date "${REMOTE_VERSIONS[$i]}" || continue # skip version
+
+    if [ "$version_date_YMD" -ge "$cuttoff_date_YMD" ]; then
+      print_action "Keeping" "${REMOTE_VERSIONS[$i]}" "$created"
+    else
+      delete_version "${REMOTE_VERSIONS[$i]}" "$created"
+    fi
+  done
+}
+
```

Comment:
The algorithm in the [archive function](https://github.com/probonopd/AppImages/pull/41/files#diff-82fd3e7970204b64763ae4548bb471c9R127) has been written to suit MuseScore's needs, but can be tweaked to suit your needs. This [gist](https://gist.github.com/shoogle/e16f4d7682948fc94c7a) shows a simulation of running it against the package [musescore/nightlies-linux/MuseScoreNightly-master-x86_64](https://bintray.com/musescore/nightlies-linux/MuseScoreNightly-master-x86_64/view#files).